operator/internal/controller/redpanda/redpanda_controller.go

```diff
@@ -112,390 +107,319 @@ type RedpandaReconciler struct {
-		log.V(logger.TraceLevel).Info(durationMsg)
-	}()
-
-	log.V(logger.TraceLevel).Info("Starting reconcile loop")
+	defer timing.Execution(ctx).Stop("reconciling redpanda cluster")
 
 	rp := &redpandav1alpha2.Redpanda{}
 	if err := r.Client.Get(ctx, req.NamespacedName, rp); err != nil {
 		return ctrl.Result{}, client.IgnoreNotFound(err)
 	}
+	rp.ManagedFields = nil // nil out our managed fields
 
-	// Examine if the object is under deletion
-	if !rp.ObjectMeta.DeletionTimestamp.IsZero() {
-		return ctrl.Result{}, r.reconcileDelete(ctx, rp)
-	}
-
-	if !isRedpandaManaged(ctx, rp) {
-		if controllerutil.ContainsFinalizer(rp, FinalizerKey) {
-			// if no longer managed by us, attempt to remove the finalizer
-			controllerutil.RemoveFinalizer(rp, FinalizerKey)
-			if err := r.Client.Update(ctx, rp); err != nil {
-				return ctrl.Result{}, errors.WithStack(err)
-			}
-		}
-
-		return ctrl.Result{}, nil
-	}
-
+	// TODO: remove this when we delete the managed decommission controller
 	_, ok := rp.GetAnnotations()[resources.ManagedDecommissionAnnotation]
 	if ok {
-		log.V(logger.TraceLevel).Info("Managed decommission")
+		log.V(TraceLevel).Info("Managed decommission")
 		return ctrl.Result{}, nil
 	}
 
-	// add finalizer if not exist
-	if !controllerutil.ContainsFinalizer(rp, FinalizerKey) {
-		patch := client.MergeFrom(rp.DeepCopy())
-		controllerutil.AddFinalizer(rp, FinalizerKey)
-		if err := r.Client.Patch(ctx, rp, patch); err != nil {
-			log.Error(err, "unable to register finalizer")
-			return ctrl.Result{}, errors.WithStack(err)
-		}
-	}
-
-	// Upgrade checks. Don't reconcile if UseFlux is true or if ChartRef is set.
-	if rp.Spec.ChartRef.UseFlux != nil && *rp.Spec.ChartRef.UseFlux {
-		log.Error(nil, "useFlux: true is no longer supported. Please downgrade or unset `useFlux`")
-		return ctrl.Result{}, nil
+	// grab our existing and desired pool resources
+	// so that we can immediately calculate cluster status
+	// from and sync in any subsequent operation that
+	// early returns
+	injectedConfigVersion := ""
+	if rp.Annotations != nil && rp.Annotations[RestartClusterOnConfigChangeKey] == "true" {
+		injectedConfigVersion = rp.Status.ConfigVersion
 	}
-
-	if rp.Spec.ChartRef.ChartVersion != "" {
-		msg := "Specifying chartVersion is no longer supported. Please downgrade or unset `chartRef.chartVersion`"
-		log.Error(nil, msg)
-		r.EventRecorder.Eventf(rp, "Warning", redpandav1alpha2.EventSeverityError, msg)
-		return ctrl.Result{}, nil
+	pools, err := r.LifecycleClient.FetchExistingAndDesiredPools(ctx, rp, injectedConfigVersion)
+	if err != nil {
+		log.Error(err, "fetching pools")
+		return ctrl.Result{}, err
 	}
 
-	if err := r.reconcileResources(ctx, rp); err != nil {
-		return ctrl.Result{}, errors.WithStack(err)
-	}
+	status := lifecycle.NewClusterStatus()
+	status.Pools = pools.PoolStatuses()
 
-	if err := r.reconcileStatus(ctx, rp); err != nil {
-		return ctrl.Result{}, errors.WithStack(err)
+	if pools.AnyReady() {
```

Comment:
We're going to want to touch base with cloud about what statuses they're using and how they're using them before merging this.